File: legacy/artifacts/src/android/utils/AndroidCodeUtils.ts
export module AndroidCodeUtils {
  export function classSignatureToFqn(classSignature: string): string {
    // Remove leading 'L' and trailing ';' if they exist
    if (classSignature.startsWith('L')) {
      classSignature = classSignature.substring(1, classSignature.length);
    }
    if (classSignature.endsWith(';')) {
      classSignature = classSignature.substring(0, classSignature.length - 1);
    }

    return classSignature.replace(/\//g, '.');
  }

  export function fqnToClassSignature(fqn: string): string {
    fqn = fqn.replace(/\./g, '/');
    return `L${fqn};`;
  }

  export function removeKotlinSuffixFromSignature(classSignature: string): string {
    return classSignature.replace(/Kt(?=[$\/;])/, '');
  }

  export function removeKotlinSuffixFromFqn(fqn: string): string {
    return fqn.replace(/Kt(?=[$.]|\b)/, '');
  }
}

File: src/launchpad/parsers/android/dex/android_code_utils.py
import re


class AndroidCodeUtils:
    @staticmethod
    def class_signature_to_fqn(class_signature: str) -> str:
        # Remove leading 'L' and trailing ';' if they exist
        if class_signature.startswith("L"):
            class_signature = class_signature[1:]
        if class_signature.endswith(";"):
            class_signature = class_signature[:-1]

        return class_signature.replace("/", ".")

    @staticmethod
    def fqn_to_class_signature(fqn: str) -> str:
        fqn = fqn.replace(".", "/")
        return f"L{fqn};"

    @staticmethod
    def remove_kotlin_suffix_from_signature(class_signature: str) -> str:
        return re.sub(r"Kt(?=[$/;])", "", class_signature)

    @staticmethod
    def remove_kotlin_suffix_from_fqn(fqn: str) -> str:
        return re.sub(r"Kt(?=[$.]|\b)", "", fqn)
